File: Android/tuilivekit/livekit/src/main/java/com/trtc/uikit/livekit/component/floatwindow/view/RoundFrameLayout.java
package com.trtc.uikit.livekit.component.floatwindow.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Region;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;

import com.tencent.qcloud.tuicore.util.ScreenUtil;

public final class RoundFrameLayout extends FrameLayout {

    private int   mRadius;
    private RectF mRect;
    private Path  mPath;

    public RoundFrameLayout(Context context) {
        this(context, null);
    }

    public RoundFrameLayout(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        setWillNotDraw(false);
        mRadius = ScreenUtil.dip2px(12);
        mPath = new Path();
    }

    public void setRadius(int radius) {
        mRadius = radius;
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        super.onSizeChanged(w, h, oldw, oldh);
        mRect = new RectF(0, 0, getWidth(), getHeight());
    }

    @Override
    public void draw(Canvas canvas) {
        mPath.reset();
        mPath.addRoundRect(mRect, mRadius, mRadius, Path.Direction.CW);
        int saveCount = canvas.save();
        canvas.clipPath(mPath, Region.Op.INTERSECT);
        super.draw(canvas);
        canvas.restoreToCount(saveCount);
    }
}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/common/ui/RoundFrameLayout.kt
package com.trtc.uikit.livekit.common.ui

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.widget.FrameLayout
import androidx.annotation.Nullable
import com.tencent.qcloud.tuicore.util.ScreenUtil

class RoundFrameLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : FrameLayout(context, attrs) {
    private var radius = ScreenUtil.dip2px(12f)
    private var rect: RectF? = null
    private val path = Path()

    init {
        setWillNotDraw(false)
    }

    fun setRadius(radius: Int) {
        this.radius = radius
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        rect = RectF(0f, 0f, w.toFloat(), h.toFloat())
    }

    override fun draw(canvas: Canvas) {
        path.reset()
        rect?.let { path.addRoundRect(it, radius.toFloat(), radius.toFloat(), Path.Direction.CW) }
        canvas.save().also { saveCount ->
            canvas.clipPath(path, Region.Op.INTERSECT)
            super.draw(canvas)
            canvas.restoreToCount(saveCount)
        }
    }
}